### PR TITLE
Add option to limit progress bar width

### PR DESCRIPTION
**Pull request by diizzyy**
This option makes progress bar not scale to column width and sets it to fit within 80 cols

### CODE REVIEW

**Comment by complexlogic** [comment]
Can you explain the motivation for this change? What issue does this solve?

**Comment by diizzyy** [comment]
Even if you have a window open that's wider than lets say 80 cols necessarily mean that you only want a progress bar to cover the full width as it reduces for readability (for example exhale also limits width). LAME is the only encoder I know that also uses full width

Looking at a few others,
flac (no progressbar but indicator)
twolame (no progressbar but indicator)
oggenc (no progressbar but indicator)
opusenc (no progressbar but indicator)
wavpack (no progressbar but indicator)
lame (no progressbar but indicator)

**Comment by complexlogic** [comment]
Thanks for the context. I can see how this feature could be desirable, but I have a few suggestions:
- I don't want the progress bar to be fixed width. That means the progress bar will wrap to the next line if the console width is less than 80 characters. I think this feature would be better implemented as a *maximum* width instead.
- Perhaps instead of an ON/OFF setting, allow the user to specify the maximum. For example, `-DMAX_PROGBAR_WIDTH=80`

I might be able to use this feature on the Windows build to increase the performance of the progress bar.

**Comment by diizzyy** [comment]
@complexlogic 
Something like that?

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,8 @@ project(rsgain
   LANGUAGES CXX
 )
 
+option(SCALEPROGBAR "Scale progress bar to column width" ON)
```
Comment (complexlogic) [inline CMakeLists.txt]: No need to have the on/off option. We can check for the existence of the width variable to determine the state of the setting

File: CMakeLists.txt [hunk 2]
```diff
@@ -12,6 +12,8 @@ project(rsgain
...
+set(PROGBARWIDTH "80" CACHE STRING "Fixed width of progress bar")
```
Comment (complexlogic) [inline CMakeLists.txt]: Should be called MAXPROGBARWIDTH to state that it is a maximum value, not fixed

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -27,6 +27,10 @@ if (WIN32)
 
 elseif (UNIX)
   add_executable(${EXECUTABLE_TITLE} ${SOURCE_FILES})
+  if(SCALEPROGBAR)
```
Comment (complexlogic) [inline src/CMakeLists.txt]: Refactor this to check for the definition of the cache variable as defined as above. You will also need to add a bounds check. i.e. the value should not be negative

File: src/output.cpp [hunk 4]
```diff
@@ -72,9 +72,13 @@ void ProgressBar::update(int pos)
 	if (pos == pos_prev)
 		return;
 
+#ifdef SCALEDPROGRESSBAR
```
Comment (complexlogic) [inline src/output.cpp]: Remove the condtional for the on/off. Always check for the width, then set to the max if it exceeds. Something like this:
```C++
w = this->get_console_width();
#ifdef MAXPROGBARWIDTH
if (w > MAXPROGBARWIDTH)
    w = MAXPROGBARWIDTH
#endif
w -= 8;
if (w <= 0)
    return;
```

**Review by complexlogic — CHANGES_REQUESTED** [review]

**Comment by diizzyy** [comment]
That seems like a more clean approach. Unfortunately I'll be busy until ~mid next week so feel free to refactor and commit if you want.

**Pull request merged**